### ISSUE TITLE
ICE in --emit mir: expected allocation ID to point to memory

### ISSUE THREAD

**Issue by bjorn3** [rust-lang/rust#66930]
Minimal reproducer:

```rust
static UTF8_CHAR_WIDTH: [u8; 0] = [];

pub fn utf8_char_width(b: u8) -> usize {
    UTF8_CHAR_WIDTH[b as usize] as usize
}
```
Make sure to run rustc with `--emit mir` to reproduce!

## original issue

```
error: internal compiler error: src/librustc/mir/interpret/mod.rs:473: expected allocation ID 337 to point to memory

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:892:9
stack backtrace:
   0: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
   1: core::fmt::write
   2: std::io::Write::write_fmt
   3: std::panicking::default_hook::{{closure}}
   4: std::panicking::default_hook
   5: rustc_driver::report_ice
   6: std::panicking::rust_panic_with_hook
   7: std::panicking::begin_panic
   8: rustc_errors::HandlerInner::bug
   9: rustc_errors::Handler::bug
  10: rustc::util::bug::opt_span_bug_fmt::{{closure}}
  11: rustc::ty::context::tls::with_opt::{{closure}}
  12: rustc::ty::context::tls::with_opt
  13: rustc::util::bug::opt_span_bug_fmt
  14: rustc::util::bug::bug_fmt
  15: <rustc::ty::print::pretty::FmtPrinter<F> as rustc::ty::print::Printer>::print_const
  16: rustc::ty::print::pretty::<impl core::fmt::Display for &rustc::ty::sty::Const>::fmt
  17: core::fmt::write
  18: core::fmt::Formatter::write_fmt
  19: <rustc::mir::Constant as core::fmt::Display>::fmt
  20: core::fmt::write
  21: core::fmt::Formatter::write_fmt
  22: <&T as core::fmt::Debug>::fmt
  23: core::fmt::write
  24: core::fmt::Formatter::write_fmt
  25: <rustc::mir::Operand as core::fmt::Debug>::fmt
  26: core::fmt::write
  27: core::fmt::Formatter::write_fmt
  28: <rustc::mir::Rvalue as core::fmt::Debug>::fmt
  29: core::fmt::write
  30: core::fmt::Formatter::write_fmt
  31: <rustc::mir::Statement as core::fmt::Debug>::fmt
  32: <&T as core::fmt::Debug>::fmt
  33: core::fmt::write
  34: alloc::fmt::format
  35: rustc_codegen_cranelift::base::trans_stmt
  36: rustc_codegen_cranelift::base::codegen_fn_content
  37: rustc_codegen_cranelift::base::trans_fn
  38: rustc_codegen_cranelift::driver::trans_mono_item
  39: rustc_codegen_cranelift::driver::codegen_mono_items::{{closure}}::{{closure}}
  40: rustc_codegen_cranelift::unimpl::try_unimpl::{{closure}}::{{closure}}
  41: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  42: std::panicking::try::do_call
  43: __rust_maybe_catch_panic
  44: std::panicking::try
  45: std::panic::catch_unwind
  46: rustc_codegen_cranelift::unimpl::try_unimpl::{{closure}}
  47: std::thread::local::LocalKey<T>::try_with
  48: std::thread::local::LocalKey<T>::with
  49: rustc_codegen_cranelift::unimpl::try_unimpl
  50: rustc_codegen_cranelift::driver::codegen_mono_items::{{closure}}
  51: rustc_codegen_cranelift::driver::time
  52: rustc_codegen_cranelift::driver::codegen_mono_items
  53: rustc_codegen_cranelift::driver::codegen_cgus
  54: rustc_codegen_cranelift::driver::run_aot
  55: rustc_codegen_cranelift::driver::codegen_crate
  56: <rustc_codegen_cranelift::CraneliftCodegenBackend as rustc_codegen_utils::codegen_backend::CodegenBackend>::codegen_crate
  57: rustc_interface::passes::start_codegen::{{closure}}
  58: rustc_interface::passes::start_codegen
  59: rustc::ty::context::tls::enter_global
  60: rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}
  61: rustc_interface::passes::create_global_ctxt::{{closure}}
  62: rustc_interface::passes::BoxedGlobalCtxt::enter
  63: rustc_interface::queries::Query<T>::compute
  64: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::ongoing_codegen
  65: rustc_interface::interface::run_compiler_in_existing_thread_pool
  66: std::thread::local::LocalKey<T>::with
  67: scoped_tls::ScopedKey<T>::set
  68: syntax::with_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.41.0-nightly (25d8a9494 2019-11-29) running on x86_64-apple-darwin

note: compiler flags: -Z panic-abort-tests -Z codegen-backend=/Users/bjorn/Documents/rustc_codegen_cranelift/target/debug/librustc_codegen_cranelift.dylib -Z force-unstable-if-unmarked -C debuginfo=2 -C incremental -C panic=abort -C debuginfo=2 --crate-type lib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
Instance { def: Item(DefId(0:6927 ~ core[c569]::str[0]::utf8_char_width[0])), substs: [] } _ZN4core3str15utf8_char_width17h59697a8cf43d24a0E
error: aborting due to previous error

error: could not compile `core`.
```

Using `-Zmir-opt-level=0` doesn't fix this.

**Comment by bjorn3** [comment]
Minimal repro:

```rust
static UTF8_CHAR_WIDTH: [u8; 0] = [];

pub fn utf8_char_width(b: u8) -> usize {
    UTF8_CHAR_WIDTH[b as usize] as usize
}
```

**Comment by bjorn3** [comment]
Other failure with code causing the above problem:

```sh
$ cat > assertion_ice.rs
static UTF8_CHAR_WIDTH: [u8; 0] = [];

pub fn utf8_char_width(b: u8) -> usize {
    UTF8_CHAR_WIDTH[b as usize] as usize
}
$ rustc +nightly assertion_ice.rs --emit mir --crate-type lib
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `0`', src/librustc/mir/interpret/value.rs:319:9
stack backtrace:
   0: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
   1: core::fmt::write
   2: std::io::Write::write_fmt
   3: std::panicking::default_hook::{{closure}}
   4: std::panicking::default_hook
   5: rustc_driver::report_ice
   6: std::panicking::rust_panic_with_hook
   7: std::panicking::continue_panic_fmt
   8: std::panicking::begin_panic_fmt
   9: rustc::mir::interpret::value::Scalar<Tag>::to_bits
  10: <rustc_mir::interpret::validity::ValidityVisitor<M> as rustc_mir::interpret::visitor::ValueVisitor<M>>::visit_value
  11: rustc_mir::interpret::validity::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::validate_operand
  12: <rustc_mir::transform::const_prop::ConstPropagator as rustc::mir::visit::MutVisitor>::visit_statement
  13: <rustc_mir::transform::const_prop::ConstProp as rustc_mir::transform::MirPass>::run_pass
  14: rustc_mir::transform::run_passes
  15: rustc_mir::transform::run_optimization_passes
  16: rustc_mir::transform::optimized_mir
  17: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::optimized_mir>::compute
  18: rustc::dep_graph::graph::DepGraph::with_task_impl
  19: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  20: rustc_mir::monomorphize::collector::collect_items_rec
  21: rustc_mir::monomorphize::collector::collect_crate_mono_items::{{closure}}
  22: rustc::util::common::time
  23: rustc_mir::monomorphize::collector::collect_crate_mono_items
  24: rustc::util::common::time
  25: rustc_mir::monomorphize::partitioning::collect_and_partition_mono_items
  26: rustc::ty::query::__query_compute::collect_and_partition_mono_items
  27: rustc::dep_graph::graph::DepGraph::with_task_impl
  28: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  29: rustc_codegen_ssa::back::symbol_export::exported_symbols_provider_local
  30: rustc::ty::query::__query_compute::exported_symbols
  31: rustc::dep_graph::graph::DepGraph::with_task_impl
  32: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  33: rustc_metadata::rmeta::encoder::EncodeContext::encode_crate_root
  34: rustc::ty::context::tls::with_context::{{closure}}
  35: rustc_metadata::rmeta::encoder::encode_metadata
  36: rustc_metadata::rmeta::decoder::cstore_impl::<impl rustc::middle::cstore::CrateStore for rustc_metadata::creader::CStore>::encode_metadata
  37: rustc::ty::context::TyCtxt::encode_metadata
  38: rustc_interface::passes::start_codegen::{{closure}}
  39: rustc_interface::passes::start_codegen
  40: rustc::ty::context::tls::enter_global
  41: rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}
  42: rustc_interface::passes::create_global_ctxt::{{closure}}
  43: rustc_interface::passes::BoxedGlobalCtxt::enter
  44: rustc_interface::queries::Query<T>::compute
  45: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::ongoing_codegen
  46: rustc_interface::interface::run_compiler_in_existing_thread_pool
  47: std::thread::local::LocalKey<T>::with
  48: scoped_tls::ScopedKey<T>::set
  49: syntax::with_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.41.0-nightly (25d8a9494 2019-11-29) running on x86_64-apple-darwin

note: compiler flags: --crate-type lib

query stack during panic:
#0 [optimized_mir] processing `utf8_char_width`
#1 [collect_and_partition_mono_items] collect_and_partition_mono_items
#2 [exported_symbols] exported_symbols
end of query stack
```

This one is observable by printing the whole mir using `--emit mir`, unlike the other, which requires you to debug print every statement.

**Comment by Centril** [comment]
cc @eddyb

**Comment by eddyb** [comment]
That second one is a miri assertion:
https://github.com/rust-lang/rust/blob/2da942f32802c8233a09744024dfbc34431adf65/src/librustc/mir/interpret/value.rs#L319
Interestingly, there's an `assert_ne!` afterwards that would be more helpful if it were before: https://github.com/rust-lang/rust/blob/2da942f32802c8233a09744024dfbc34431adf65/src/librustc/mir/interpret/value.rs#L320

cc @oli-obk

**Comment by JohnTitor** [comment]
It seems like no longer ICE with latest nightly, I guess #67501 fixed this (CC: @oli-obk).
Marked as E-needstest

**Comment by JohnTitor** [comment]
@Alexendoo points out [the second example](https://github.com/rust-lang/rust/issues/66930#issuecomment-560097226) hasn't been fixed yet, dropping E-needstest.

**Comment by RalfJung** [comment]
Is the issue description still accurate? "Can't debug format a mir stmt in DefId" is an assertion I added after we fixed a long-standing ICE in MIR debug printing... and I can't even find the assertion any more, maybe it got removed. :(

The other ICEs described here sound very different than that, they look like CTFE is not handling ZSTs properly somewhere.

**Comment by RalfJung** [comment]
> @Alexendoo points out the second example hasn't been fixed yet, dropping E-needstest.

I cannot [reproduce the ICE](https://play.rust-lang.org/?version=beta&mode=debug&edition=2018&gist=15a57328a6ddb5d857b49b52a3b3da49) on beta or nightly, so it looks pretty fixed to me.

@bjorn3 can you still reproduce this issue with latest nightly? The first backtrace has no code, so it is impossible to debug; the second backtrace (the one with self-contained code) does not seem to ICE any more.

**Comment by JohnTitor** [comment]
> I cannot reproduce the ICE on beta or nightly, so it looks pretty fixed to me.

Uhm? I don't think so, did you run with `--emit=mir`?

**Comment by RalfJung** [comment]
Ah, I just ran on playground.

It would be good to have up-to-date reproduction instructions in the first post, and not just in the third post *below* a huge stacktrace.

**Comment by JohnTitor** [comment]
> It would be good to have up-to-date reproduction instructions in the first post, and not just in the third post below a huge stacktrace.

Yeah, but I don't have write access :/

**Comment by RalfJung** [comment]
> Yeah, but I don't have write access :/

Ouch, sorry. :( I didn't know.
I will update the OP.

**Comment by RalfJung** [comment]
<details><summary>Here's a backtrace with debug info</summary>

```
thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:881:9
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /home/r/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /home/r/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1053
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1428
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:204
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:224
  10: <alloc::boxed::Box<F> as core::ops::function::Fn<A>>::call
             at ./src/liballoc/boxed.rs:1031
  11: rustc_driver::report_ice
             at src/librustc_driver/lib.rs:1179
  12: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:474
  13: std::panicking::begin_panic
             at ./src/libstd/panicking.rs:397
  14: rustc_errors::HandlerInner::bug
             at src/librustc_errors/lib.rs:881
  15: rustc_errors::Handler::bug
             at src/librustc_errors/lib.rs:656
  16: rustc::util::bug::opt_span_bug_fmt::{{closure}}
             at src/librustc/util/bug.rs:36
  17: rustc::ty::context::tls::with_opt::{{closure}}
             at src/librustc/ty/context.rs:1829
  18: rustc::ty::context::tls::with_context_opt
             at src/librustc/ty/context.rs:1781
  19: rustc::ty::context::tls::with_opt
             at src/librustc/ty/context.rs:1829
  20: rustc::util::bug::opt_span_bug_fmt
             at src/librustc/util/bug.rs:32
  21: rustc::util::bug::bug_fmt
             at src/librustc/util/bug.rs:12
  22: rustc::mir::interpret::AllocMap::unwrap_memory
             at src/librustc/mir/interpret/mod.rs:482
  23: rustc::ty::print::pretty::PrettyPrinter::pretty_print_const_value
             at src/librustc/ty/print/pretty.rs:963
  24: rustc::ty::print::pretty::PrettyPrinter::pretty_print_const
             at src/librustc/ty/print/pretty.rs:882
  25: <rustc::ty::print::pretty::FmtPrinter<F> as rustc::ty::print::Printer>::print_const
             at src/librustc/ty/print/pretty.rs:1184
  26: <&rustc::ty::sty::Const as rustc::ty::print::Print<P>>::print
             at src/librustc/ty/print/mod.rs:345
  27: rustc::ty::print::pretty::<impl core::fmt::Display for &rustc::ty::sty::Const>::fmt::{{closure}}
             at src/librustc/ty/print/pretty.rs:1624
  28: rustc::ty::context::tls::with::{{closure}}
             at src/librustc/ty/context.rs:1819
  29: rustc::ty::context::tls::with_context::{{closure}}
             at src/librustc/ty/context.rs:1792
  30: rustc::ty::context::tls::with_context_opt
             at src/librustc/ty/context.rs:1781
  31: rustc::ty::context::tls::with_context
             at src/librustc/ty/context.rs:1792
  32: rustc::ty::context::tls::with
             at src/librustc/ty/context.rs:1819
  33: rustc::ty::print::pretty::<impl core::fmt::Display for &rustc::ty::sty::Const>::fmt
             at src/librustc/ty/print/pretty.rs:1623
  34: core::fmt::write
             at src/libcore/fmt/mod.rs:1053
  35: core::fmt::Formatter::write_fmt
             at src/libcore/fmt/mod.rs:1474
  36: <rustc::mir::Constant as core::fmt::Display>::fmt
             at ./src/libcore/fmt/mod.rs:0
  37: core::fmt::write
             at src/libcore/fmt/mod.rs:1053
  38: core::fmt::Formatter::write_fmt
             at src/libcore/fmt/mod.rs:1474
  39: <rustc::mir::Constant as core::fmt::Debug>::fmt
             at src/librustc/mir/mod.rs:2559
  40: core::fmt::write
             at src/libcore/fmt/mod.rs:1053
  41: core::fmt::Formatter::write_fmt
             at src/libcore/fmt/mod.rs:1474
  42: <rustc::mir::Operand as core::fmt::Debug>::fmt
             at src/librustc/mir/mod.rs:0
  43: core::fmt::write
             at src/libcore/fmt/mod.rs:1053
  44: core::fmt::Formatter::write_fmt
             at src/libcore/fmt/mod.rs:1474
  45: <rustc::mir::Rvalue as core::fmt::Debug>::fmt
             at src/librustc/mir/mod.rs:0
  46: core::fmt::write
             at src/libcore/fmt/mod.rs:1053
  47: core::fmt::Formatter::write_fmt
             at src/libcore/fmt/mod.rs:1474
  48: <rustc::mir::Statement as core::fmt::Debug>::fmt
             at src/librustc/mir/mod.rs:0
  49: core::fmt::run
             at src/libcore/fmt/mod.rs:1087
  50: core::fmt::write
             at src/libcore/fmt/mod.rs:1062
  51: core::fmt::Write::write_fmt
             at ./src/libcore/fmt/mod.rs:195
  52: alloc::fmt::format
             at src/liballoc/fmt.rs:571
  53: rustc_mir::util::pretty::write_basic_block
             at src/librustc_mir/util/pretty.rs:303
  54: rustc_mir::util::pretty::write_mir_fn
             at src/librustc_mir/util/pretty.rs:272
  55: rustc_mir::util::pretty::write_mir_pretty
             at src/librustc_mir/util/pretty.rs:248
  56: rustc_mir::transform::dump_mir::emit_mir
             at src/librustc_mir/transform/dump_mir.rs:65
  57: rustc_interface::passes::start_codegen
             at src/librustc_interface/passes.rs:973
  58: rustc_interface::queries::Queries::ongoing_codegen::{{closure}}::{{closure}}
             at src/librustc_interface/queries.rs:280
  59: rustc_interface::passes::QueryContext::enter::{{closure}}
             at src/librustc_interface/passes.rs:696
  60: rustc::ty::context::tls::enter_global::{{closure}}
             at ./src/librustc/ty/context.rs:1743
  61: rustc::ty::context::tls::enter_context::{{closure}}
             at ./src/librustc/ty/context.rs:1720
  62: rustc::ty::context::tls::set_tlv
             at ./src/librustc/ty/context.rs:1704
  63: rustc::ty::context::tls::enter_context
             at ./src/librustc/ty/context.rs:1720
  64: rustc::ty::context::tls::enter_global
             at ./src/librustc/ty/context.rs:1743
  65: rustc_interface::passes::QueryContext::enter
             at src/librustc_interface/passes.rs:696
  66: rustc_interface::queries::Queries::ongoing_codegen::{{closure}}
             at src/librustc_interface/queries.rs:274
  67: rustc_interface::queries::Query<T>::compute
             at src/librustc_interface/queries.rs:33
  68: rustc_interface::queries::Queries::ongoing_codegen
             at src/librustc_interface/queries.rs:272
  69: rustc_driver::run_compiler::{{closure}}::{{closure}}
             at src/librustc_driver/lib.rs:399
  70: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
             at ./src/librustc_interface/queries.rs:339
  71: rustc_driver::run_compiler::{{closure}}
             at src/librustc_driver/lib.rs:289
  72: rustc_interface::interface::run_compiler_in_existing_thread_pool
             at ./src/librustc_interface/interface.rs:199
  73: rustc_interface::interface::run_compiler::{{closure}}
             at ./src/librustc_interface/interface.rs:213
  74: rustc_interface::util::spawn_thread_pool::{{closure}}::{{closure}}::{{closure}}
             at ./src/librustc_interface/util.rs:155
  75: scoped_tls::ScopedKey<T>::set
             at /home/r/.cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-1.0.0/src/lib.rs:137
  76: rustc_interface::util::spawn_thread_pool::{{closure}}::{{closure}}
             at ./src/librustc_interface/util.rs:151
  77: scoped_tls::ScopedKey<T>::set
             at /home/r/.cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-1.0.0/src/lib.rs:137
  78: syntax::attr::with_globals::{{closure}}
             at ./src/libsyntax/attr/mod.rs:44
  79: scoped_tls::ScopedKey<T>::set
             at /home/r/.cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-1.0.0/src/lib.rs:137
  80: syntax::attr::with_globals
             at ./src/libsyntax/attr/mod.rs:44
  81: rustc_interface::util::spawn_thread_pool::{{closure}}
             at ./src/librustc_interface/util.rs:150
  82: rustc_interface::util::scoped_thread::{{closure}}
             at ./src/librustc_interface/util.rs:125
```

</details>

@oli-obk Looks like the [`unwrap_memory` here](https://github.com/rust-lang/rust/blob/bfc32dd106f0fe191b8bcd9b76348a8875c30a60/src/librustc/ty/print/pretty.rs#L966) is wrong.

**Comment by JohnTitor** [comment]
Triage: I think that ICE is fixed by #72424, no longer occurs with the latest nightly. Marking as E-needs-test.
